Refactor Object Store Selection Modals UI

itisAliRH opened the pull request:
This pull request introduces a redesign of the Object Store Selection modals for User Preferred, Current History, Tool form and Workflow run form, aiming to enhance user experience through a consistent and modernised interface.

Required: #19521

https://github.com/user-attachments/assets/e85fca15-8c1b-4683-83bd-eb53fb2a28f7

|Before|After|
|--|--|
|<img width="960" alt="image" src="https://github.com/user-attachments/assets/65b1fabd-642a-479c-b6ac-04a249103e5c" />|<img width="960" alt="image" src="https://github.com/user-attachments/assets/978a63e6-e988-421a-a11c-37dd389e53a0" />|
|<img width="960" alt="image" src="https://github.com/user-attachments/assets/fdb6c7d8-5a2a-4ee2-899c-ce219c7cf117" />|<img width="960" alt="image" src="https://github.com/user-attachments/assets/382d59b8-c3c0-4508-abc3-762a81815140" />|
|<img width="960" alt="image" src="https://github.com/user-attachments/assets/c51a6357-4ae1-43fb-bbc1-ac59c21a8943" />|<img width="960" alt="image" src="https://github.com/user-attachments/assets/0e6faaeb-cac8-4a5f-ae98-39a1d890c185" />|
|<img width="960" alt="image" src="https://github.com/user-attachments/assets/4fcf4e35-870d-4bcd-91f6-5264cbe107e4" />|<img width="960" alt="image" src="https://github.com/user-attachments/assets/f23951f8-52e6-493f-b884-8513dd0a4e2f" />|

Key Changes:

- Enhanced Selection Interface using the `SourceOptionCard`
- Update types and improve the logic
- Use a consistent object store selection for every use case

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  a1. Go to `User Preferences`
  a2. Click on `Preferred Storage Location`

  b1. Go to the tool run form
  b2. Select the storage icon at the top of the form toolbar

  c1. Go to the workflow run form
  c2. Click on the storage icon at the top of the page

  d1. In the current history panel, click on the storage icon on top

 
## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

itisAliRH committed:
Adds headers to object store and file source routes

Includes headers "Storage Locations" and "Remote File Sources"
to the props of the respective routes for better clarity.

itisAliRH committed:
Adds headers to object store and file source routes

Includes headers "Storage Locations" and "Remote File Sources"
to the props of the respective routes for better clarity.

itisAliRH committed:
Refactors UserPreferredObjectStore component to use Vue 3 Composition API

Migrates from Vue 2 Options API to Vue 3 Composition API for better maintainability and readability. Replaces Vue instance methods and lifecycle hooks with setup function, refs, and computed properties. Updates modal and font icon usage to align with new component structure.

Fixes layout and style issues with the modal dialog.

itisAliRH committed:
Enhances SourceOptionCard UI and logic

Adds computed properties for uniqueId, buttonTitle, and buttonTooltip
Introduces optional props 'selected', 'submitButtonTooltip', and 'selectionMode'
Updates styles for selected state
Implements emit for 'select' event

Improves user interaction and flexibility of SourceOptionCard component

itisAliRH committed:
Refactors SelectObjectStore component layout and logic

Replaces button-based object store selection with SourceOptionCard components
Removes unused imports and refactors prop types to use lowercase
Simplifies handleSubmit function and error handling

Enhances user interface for better usability and maintainability

itisAliRH committed:
Export UserConcreteObjectStoreModel type

Export type definition for UserConcreteObjectStoreModel in API schema handling.

itisAliRH committed:
Refactors SourceOption type and uniqueId computation

Replaces ConcreteObjectStoreModel with UserConcreteObjectStoreModel in SourceOption type
Enhances uniqueId computation to handle object_store_id
Adds data attribute to source option card for easier identification

itisAliRH committed:
Updates object store type and loading logic

Replaces ConcreteObjectStoreModel with UserConcreteObjectStoreModel.
Simplifies isLoaded computed property logic.
Typecasts loaded data to ensure correct type.

itisAliRH committed:
Updates object store type and default value

Replaces ConcreteObjectStoreModel with UserConcreteObjectStoreModel.
Adds new fields to default object store configuration because of the new type

itisAliRH committed:
Updates object store selectors

Replaces option buttons with modal and option cards in object store selection.
Refactors schema to align with new UI structure.

itisAliRH committed:
Migrates test file to TypeScript and refactors selectors

Converts `UserPreferredObjectStore.test.js` to TypeScript, updates import paths, and refactors component and selector names for consistency and clarity.

Simplifies test logic by removing obsolete constants and correcting assertions.

itisAliRH committed:
Handles null object store ID

Updates handleSubmit function to check for "__null__" ID and set it to null.

jmchilton commented on client/src/components/ObjectStore/SelectObjectStore.vue:
```diff
@@ -63,57 +45,72 @@ const error = computed(() => {
         <div v-else>
-            <b-alert v-if="error" variant="danger" class="object-store-selection-error" show>
+            <span v-localize>
+                Select a preferred storage location for new datasets. Depending on the job and workflow execution
```
Can you make the extra text dependent on the option again? I think adding that option in was a solid usability enhancement for main.

itisAliRH committed:
Removes redundant state management

Eliminates unnecessary "selectedObjectStoreId" ref
Uses prop "toolPreferredObjectStoreId" directly in template
Simplifies handleSubmit function

itisAliRH committed:
Removes unnecessary code in ToolCard component

Eliminates redundant property binding and modal visibility toggle in ToolCard component to simplify the code and improve maintainability.

itisAliRH committed:
Updates object store selection test selectors

Replaces selectors for object store option buttons with option cards

itisAliRH committed:
Updates object store selection test selectors

Modifies test cases to use `option_cards` and `option_card_select` selectors instead of `option_buttons` and `option_button`, ensuring compatibility with the updated component structure.

itisAliRH committed:
Removes unnecessary modal close logic

Eliminates redundant line that closes the preferred object store modal
to simplify the onUpdatePreferredObjectStoreId function

itisAliRH committed:
Adds condition to display storage location preference

Integrates storage location configuration to conditionally display preferred storage location message.

ref: https://github.com/galaxyproject/galaxy/pull/19697/files/78af264176fb3f726376c5a40776d2542720dff9#r1970123333

itisAliRH committed:
Adds quota usage display to SourceOptionCard

Integrates quota usage data and loading spinner to SourceOptionCard component.
Imports necessary components and computes quota source label.

ElectronicBlueberry committed:
move and refactor storage selector

itisAliRH committed:
Adds tooltip to storage location button and cancel variant to modal

Introduces a non-interactive hover tooltip for the storage location button to improve user guidance. Adds an outline-primary variant to the cancel button in the modal for better visual consistency.

itisAliRH committed:
Adds scrollable and centered properties to modals

Enhances user experience by making modals scrollable and centered
across various components for constancy.

itisAliRH committed:
Improves storage location modal handling

Integrates confirmation dialog for setting private datasets
Replaces manual modal control with prop-based visibility
Simplifies modal reset logic and emits close event

itisAliRH committed:
Updates preferred store selection tests

Refactors mountComponent to accept preferredObjectStoreId parameter.
Adds API mock for user usage.
Includes stubs for BModal to simulate modal interactions.
Updates test cases to verify modal interactions.

itisAliRH committed:
Updates object store selection element

Changes selectors from option buttons to option cards in the test to align with updated UI components for object store selection.

itisAliRH committed:
Updates storage selection method

Replaces option_buttons with option_cards for storage selection
Adds check and click for primary button if present

itisAliRH committed:
Fix type in addOrUpdateObjectStore function parameter

itisAliRH committed:
Fixes conditional rendering for icon display

Adds a condition to check for the presence of an icon before rendering it in the button component, preventing error.

itisAliRH committed:
Updates object store card selectors

Revises selectors for object store cards to use g-card ids and more specific

itisAliRH committed:
Refines UI behavior and loading indicators in source card

Replaces Bootstrap spinner with a custom loading component for better UX consistency. Improves button tooltips and titles to provide clearer messaging in selection mode.
Refactors primary actions into a computed property to enhance reactivity.
Adjusts template structure to render quota.

itisAliRH committed:
Refactors SelectObjectStore component

Reorganizes props by using the TypScript interfaces and adding default  values using `withDefaults` for better clarity and consistency.
Hide quota for default object store.
Simplifies UI text by using verbose dynamic `forWhat` prop value.
Removes redundant type annotations for
cleaner code.

itisAliRH committed:
Improves current history object store selection

- Refactors modal logic to ensure proper event handling and reset behavior.
- Simplifies error state initialization.
- Adjusts modal actions to prevent unintended closure.
- Enhances UI consistency by updating modal properties like title tags and button states.

Improves user experience by refining modal usability and ensuring accurate state management.

itisAliRH committed:
Refactors ToolSelectPreferredObjectStore and streamlines event handling

- Simplifies the props definition by removing default values and renaming the interface for clarity.
- Streamlines event handling by reordering and grouping related declarations for improved readability.

itisAliRH committed:
Convert ToolSelectPreferredObjectStore test file to TypeScript and improve test logic

- Renames a test file from JavaScript to TypeScript for type safety and consistency.
- Updates test logic to verify object store selection behavior.

itisAliRH committed:
Refactor ToolCard modal and icons

- Switches from using a font-awesome span element to the FontAwesomeIcon component for improved readability and maintainability.
- Updates modal properties to use `BModal` with enhanced configuration, including `ok-only` and `ok-title` and adjusts modal size to `lg` and modifies the title and footer options for better user experience and consistency.

itisAliRH committed:
Updates UserPreferredObjectStore modal attributes for consistent UX

- Changes the modal title tag from h2 to h3 for semantic consistency.
- Adds an "ok-only" button with a custom "Close" label.

itisAliRH committed:
Enhances WorkflowStorageConfiguration modals UX

- Adds usability improvements to modals by making them larger, scrollable, and centered.
- Add "ok-only" button
with a "Close" label for better clarity.
- Removes unnecessary state toggling in update methods for consistency.

itisAliRH committed:
Simplifies object store selection by removing redundant state

- Removes the unnecessary `selectedObjectStoreId` state and directly uses `invocationPreferredObjectStoreId` from props for binding.

itisAliRH committed:
Converts test file to TypeScript and refines test logic

- Convert the test file from JavaScript to TypeScript for improved type safety.
- Updates import paths and resolves schema reference adjustments.
- Enhances test logic by refining object store selection checks and emitted event validation.

itisAliRH committed:
Update description rendering

- Render full description.
- Updates the description binding to directly use the `sourceOption.description` property, reducing dependency on external computations.

davelopez reviewed (approved):
Works well and looks great on my tests!

Thank you @itisAliRH!

mvdbeek commented on client/src/components/ConfigTemplates/SourceOptionCard.vue:
```diff
@@ -17,40 +23,84 @@ type OptionType = {
+    } else if ("object_store_id" in props.sourceOption && props.sourceOption.object_store_id) {
+        return props.sourceOption.object_store_id;
+    } else {
+        return Math.random().toString(36).substring(7);
```
Can you add a comment here to explain what you're doing here and when you wouldn't have an id ?

itisAliRH replied:
@mvdbeek The `sourceOption` has an id if it's an instance of `FileSourceTemplateSummary` or `ObjectStoreTemplateSummary`, but `object_store_id` can be `undefined` in `UserConcreteObjectStoreModel`.
This `id` used to be required before [this change](https://github.com/galaxyproject/galaxy/pull/20064/files#diff-5e6ecdc70b61f43fa634318c28f367814cdb95867d48602f572bd8d67612d43bL19-R21), but now `GCard` auto-generates it if it’s missing, so this line can be removed. Thanks for pointing this out!

itisAliRH committed:
Replaces random ID generation with undefined fallback

Updates the computed `uniqueId` to return `undefined` instead of a random string when no valid sourceOption ID is available.
GCard assign an id automatically.
Adds id for mocked source options.

itisAliRH committed:
Adds confirmation button handling for object store selection

Enhances the object store selection process by introducing logic to handle a confirmation button.

itisAliRH committed:
Fix object store button selection

itisAliRH committed:
Remove redundant primary button click in object store selection process

itisAliRH committed:
Improve object store selection button handling to check for absence before clicking

itisAliRH committed:
Adds ID to modal for tools preferred object store selection

davelopez reviewed (approved):
Thank you @itisAliRH!
Let's get this in early so people can try it, and then we can iterate if needed during the dev cycle 👍
